// kitsune-api/lib/anime/anime.js
const axios = require('axios')
const cheerio = require('cheerio')
const author = "@Iqbalzz"

function konachan(chara){
	return new Promise((resolve,reject) => {
		let text = chara.replace(' ','_')
		axios.get('https://konachan.net/post?tags='+ text +'+')
		.then(({ data }) =>{
		const $$ = cheerio.load(data)
		const no = [];
		$$('div.pagination > a').each(function(c,d){
			no.push($$(d).text())
		})
		let mat = Math.floor(Math.random() * no.length)
		axios.get('https://konachan.net/post?page='+ mat +'&tags=' + text + '+')
		.then(({ data }) => {
		const $ = cheerio.load(data)
	 const result = [];
		$('#post-list > div.content > div:nth-child(4) > ul > li > a.directlink.largeimg').each(function(a,b){
			result.push($(b).attr('href'))
		})
	 resolve(result)
		})
	})
		.catch(reject)
	})
}

const anime = (query) => {
    return new Promise((resolve, reject) => {
        axios.get(`https://www.anime-planet.com/anime/all?name=${query}`)
            .then(({
                data
            }) => {
                const hasil = []
                const $ = cheerio.load(data)
                $('#siteContainer > ul.cardDeck.cardGrid > li ').each(function (a, b) {
                        result = {
                            status: 200,
                            author: author,
                            judul: $(b).find('> a > h3').text(),
                            link: 'https://www.anime-planet.com' + $(b).find('> a').attr('href'),
                            thumbnail: 'https://www.anime-planet.com' + $(b).find('> a > div.crop > img').attr('src')
                        };
                        hasil.push(result);
                    });
                resolve(hasil)
            })
            .catch(reject)
    })
}
const manga = (query) => {
    return new Promise((resolve, reject) => {
        axios.get(`https://www.anime-planet.com/manga/all?name=${query}`)
            .then(({
                data
            }) => {
                const hasil = []
                const $ = cheerio.load(data)
                $('#siteContainer > ul.cardDeck.cardGrid > li ').each(function (a, b) {
                        result = {
                            status: 200,
                            author: author,
                            judul: $(b).find('> a > h3').text(),
                            link: 'https://www.anime-planet.com' + $(b).find('> a').attr('href'),
                            thumbnail: 'https://www.anime-planet.com' + $(b).find('> a > div.crop > img').attr('src')
                        };
                        hasil.push(result);
                    });
                resolve(hasil)
            })
            .catch(reject)
    })
}
const characters = (query) => {
    return new Promise((resolve, reject) => {
        axios.get(`https://www.anime-planet.com/characters/all?name=${query}`)
            .then(({
                data
            }) => {
                const hasil = []
                const $ = cheerio.load(data)
                $('#siteContainer > table > tbody > tr').each(function (a, b) {
                        result = {
                            status: 200,
                            author: author,
                            character: $(b).find('> td.tableCharInfo > a').text(),
                            link: 'https://www.anime-planet.com' + $(b).find('> td.tableCharInfo > a').attr('href'),
                            thumbnail: $(b).find('> td.tableAvatar > a > img').attr('src').startsWith('https://') ? $(b).find('> td.tableAvatar > a > img').attr('src') : 'https://www.anime.planet.com' + $(b).find('> td.tableAvatar > a > img').attr('src')
                        };
                        hasil.push(result);
                    });
                resolve(hasil)
            })
            .catch(reject)
    })
}

function myanimelist (query) {
return new Promise((resolve, reject) => {
axios.get('https://myanimelist.net/anime.php?q=' + query + '&cat=anime', {
headers: {
'User-Agent': 'Mozile/5.0 (Linux; Android 9; Redmi 7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.77 Mobile Safari/537.36'
}
})
.then(({ data }) => {
const title = []
const thumb = []
const rating = []
const desc = []
const url = []
const $ = cheerio.load(data)
$('ul.data.flex > li.title').get().map((rest) => {
title.push($(rest).text())
})
$('div.icon-thumb.users.list.no-ti > img').get().map((rest) => {
thumb.push($(rest).attr('data-src'))
})
$('li.fn-grey5.fs12.fw-n').get().map((rest) => {
rating.push($(rest).text().replace('Score', ''))
})
$('li.text.mt8').get().map((rest) => {
desc.push($(rest).text())
})
$('div.box-unit1 > a').get().map((rest) => {
url.push($(rest).attr('href'))
})
const result = []
for (let i = 0; i < url.length; i++) {
result.push({
title: title[i],
thumb: thumb[i],
rating: rating[i],
desc: desc[i],
url: url[i]
})
}
resolve({
author: author,
status: true,
result: result
})
}).catch(reject)
})
}

function quotesanime() {
    return new Promise((resolve, reject) => {
        const page = Math.floor(Math.random() * 184)
        axios.get('https://otakotaku.com/quote/feed/'+page)
        .then(({ data }) => {
            const $ = cheerio.load(data)
            const hasil = []
            $('div.kotodama-list').each(function(l, h) {
                hasil.push({
                    link: $(h).find('a').attr('href'),
                    gambar: $(h).find('img').attr('data-src'),
                    karakter: $(h).find('div.char-name').text().trim(),
                    anime: $(h).find('div.anime-title').text().trim(),
                    episode: $(h).find('div.meta').text(),
                    up_at: $(h).find('small.meta').text(),
                    quotes: $(h).find('div.quote').text().trim()
                })
            })
            resolve(hasil)
        }).catch(reject)
    })
}

function anoboy(query){ //Fix Error By Iqbalzz
	return new Promise((resolve, reject) => {
		axios.get('https://anoboy.digital/?s=' + query)
			.then(({
				data
			}) => {
				const $ = cheerio.load(data)
				const format = [];
				const link = [];
				const judul = [];
				const thumb = [];
				const uptime = [];
				$('body > div.wrap > div.container > div.column-content > a > div > div.amvj > h3').each(function(a, b) {
					jud = $(b).text();
					judul.push(jud)
				})
				$('body > div.wrap > div.container > div.column-content > a > div > div.jamup').each(function(c, d) {
					upt = $(d).text();
					uptime.push(upt)
				})
				$('body > div.wrap > div.container > div.column-content > a > div > amp-img').each(function(e, f) {
					thumb.push($(f).attr('src'))
				})
				$('body > div.wrap > div.container > div.column-content > a').each(function(g, h) {
					link.push($(h).attr('href'))
				})
				for (let i = 0; i < link.length; i++) {
					format.push({
						judul: judul[i],
						thumb: thumb[i],
						link: link[i]
					})
				}
				const result = {
					creator: '@Iqbalzz',
					data: format
				}
				resolve(result)
			})
			.catch(reject)
	})
}

module.exports.myanimelist = myanimelist
module.exports.anime = anime
module.exports.manga = manga
module.exports.characters = characters
module.exports.konachan = konachan
module.exports.quotesanime = quotesanime
module.exports.anoboy = anoboy
